Atomic execution protocol and primitives

adlrocha opened the pull request:
This PR:
- [x] Implements all SCA function to handle atomic execution.
- [x] Implements the basic atomic primitives and types required for the implementation of the SCA functions.

In the next PR:
- We'll finish writing all the required primtives to implement actors that support atomic executions (i.e. `hierarchical_sca/src/atomic/mod.rs`).
- We'll write a sample actor (potentially in another repo) that use these primitives to support atomic executions.

adlrocha commented:
@aakoshh, may I have a quick review? This PR only includes the SCA functions to orchestrate atomic executions. I will work in the primitives and a sample actor in the next PR (to avoid making this one too long). Thanks :pray:

aakoshh commented on actors/hierarchical_sca/src/atomic/mod.rs:
```diff
@@ -0,0 +1,150 @@
+/// prepare the actor state for the execution, and for the merging of the
+/// output of the execution to the original state of the actor.
+pub trait LockableState<S: Serialize + DeserializeOwned> {
+    /// Merge a locked state (not necessarily the output) to the current state.
+    fn merge(other: Self) -> anyhow::Result<()>;
+    /// Merge the output of an execution to the current state.
+    fn merge_output(other: Self) -> anyhow::Result<()>;
```
```suggestion
    /// Merge a locked state (not necessarily the output) to the current state.
    fn merge(&mut self, other: Self) -> anyhow::Result<()>;
    /// Merge the output of an execution to the current state.
    fn merge_output(&mut self, other: Self) -> anyhow::Result<()>;
```
It's definitely missing an argument, and judging by the return value it wants to mutate it.

There are other alternatives that would be more functional:
```
    /// Merge a locked state (not necessarily the output) to the current state.
    fn merge(self, other: Self) -> anyhow::Result<Self>;
    /// Merge the output of an execution to the current state.
    fn merge_output(self, other: Self) -> anyhow::Result<Self>;
```

Another consideration would be whether to return the `other` as part of the error, or always consider it consumed even if it failed to merge.

adlrocha replied:
Good call! Thank you for also reviewing this file, actually these were just placeholders for the next PR where I'll add all the low-level primitives for the atomic execution, but I will fix it right away. For now I'll go with the first option (internal mutation), but it'll probably change in the next PR (once we have also a sample actor to play with and see how usable the trait is).

aakoshh commented on actors/hierarchical_sca/src/atomic/mod.rs:
```diff
@@ -0,0 +1,150 @@
+where
+    T: Serialize + DeserializeOwned + LockableState<T>,
+{
+    /// Map with all the locked state in the actor uniquely identified through
+    /// their Cid.
+    fn locked_map_cid() -> TCid<THamt<Cid, LockedState<T>>>;
+    /// Returns the output state of an execution from the current state
+    /// of the actor according to the input parameters.
+    fn output(params: LockParams) -> LockedState<T>;
```
```suggestion
    /// Map with all the locked state in the actor uniquely identified through
    /// their Cid.
    fn locked_map_cid(&self) -> TCid<THamt<Cid, LockedState<T>>>;
    /// Returns the output state of an execution from the current state
    /// of the actor according to the input parameters.
    fn output(&self, params: LockParams) -> LockedState<T>;
```
At least I'd think it's not some kind of type level function.

aakoshh replied:
Sorry, I might just be biased by my earlier experience with Actors (mainly the Akka project in Scala) being individuals. I seem to recall that actors in Filecoin were unit structs. I'll see how this works out further down.

adlrocha replied:
No, you are right. Again, this is just me copy-pasting from the Go interface and leave it as a placeholder for now. Thanks :pray:

aakoshh commented on actors/hierarchical_sca/src/atomic/mod.rs:
```diff
@@ -0,0 +1,150 @@
+    /// involved in the execution.
+    fn unlock(params: UnlockParams) -> anyhow::Result<Option<RawBytes>>;
+    /// Aborts the execution and unlocks the locked state.
+    fn abort(params: LockParams) -> anyhow::Result<Option<RawBytes>>;
```
What is the optional return value for?

adlrocha replied:
This trait is supposed to be used by a user-defined FVM actor, and for now (until there's a clearer SDK for the FVM), we are returning `Result<Option<RawBytes>>` in actor functions to signal if the have any return value or `None`. See https://github.com/adlrocha/hc-subnet-actor/blob/53627c07b21d2ae96831a94a33f274b5d53e1bc0/src/lib.rs#L42-L49

aakoshh replied:
Ah, right, I see. Maybe a simple alias like `type ActorResult = anyhow::Result<Option<RawBytes>>` would help clear it up.

So the idea is that your library code will provide the actor `invoke` method, and the developer implements this interface, and they may or may not want to return some value, which can perhaps be picked up from the logs.

adlrocha replied:
Correct. I added an alias for clarity.

aakoshh commented on actors/hierarchical_sca/src/atomic/mod.rs:
```diff
@@ -0,0 +1,150 @@
+    }
+}
+
+/// State of an actor including a lock to support atomic executions.
+#[derive(Serialize_tuple, Deserialize_tuple)]
+pub struct LockedState<T>
+where
+    T: Serialize + DeserializeOwned + LockableState<T>,
+{
+    lock: bool,
+    state: T,
```
To be honest, given that this has a `lock: bool` field, the "locked state" name is a bit misleading. Is it locked if it contains `lock: flase`? Maybe this could be called `LockableState` and the other trait `MergeableState`?

aakoshh commented on actors/hierarchical_sca/src/atomic/mod.rs:
```diff
@@ -0,0 +1,150 @@
+    fn lock(params: LockParams) -> anyhow::Result<Option<RawBytes>>;
+    /// Merges some state to the current state of the actor to prepare for the execution
+    /// of the protocol.
+    fn merge(params: MergeParams<T>) -> anyhow::Result<Option<RawBytes>>;
```
The `merge` function above doesn't have a return value. What does the `Option<RawByte>` represent here?

adlrocha replied:
See https://github.com/adlrocha/builtin-actors/pull/13#discussion_r928915526. I will add a note in the code so that it is clear that this is the trait for a lockable actor and thus the functions has the signature of an actor function.

aakoshh commented on actors/hierarchical_sca/src/atomic/mod.rs:
```diff
@@ -0,0 +1,150 @@
+    S: Serialize + DeserializeOwned + LockableActorState<T>,
+{
+    /// Locks the state to perform the execution determined by the locking params.
+    fn lock(params: LockParams) -> anyhow::Result<Option<RawBytes>>;
```
When does it return `Some` vs `None`?

adlrocha replied:
See comment above.

aakoshh commented on actors/hierarchical_sca/src/cross.rs:
```diff
@@ -80,14 +73,7 @@ impl StorableMsg {
-            value: value,
-            nonce: 0,
-        })
+        Ok(Self { from, to, method: METHOD_SEND, params: RawBytes::default(), value, nonce: 0 })
```
So top-down messages always go with 0 nonce?

adlrocha replied:
Not really. Here we are just creating a "factory message" for a fund message, that is conveniently updated when committing the cross-net message for propagation/execution. See https://github.com/adlrocha/builtin-actors/blob/9ede48182f541b565a63a294f3fc67a423128479/actors/hierarchical_sca/src/state.rs#L409

aakoshh replied:
Aha, thanks! It would be helpful to have a factory method that doesn't take a nonce, where it can be explained for good that this nonce is going to be set later.

adlrocha replied:
Fixed. Added a default for `StorableMsg` and a note on how the nonce is set for top-down messages.

aakoshh commented on actors/hierarchical_sca/src/exec.rs:
```diff
@@ -0,0 +1,197 @@
+#[repr(u64)]
+pub enum ExecStatus {
+    /// Oh, oh! something went wrong. This state should never appear.
+    UndefState,
```
Sounds more like an `Error` than a state. Can we make illegal states unrepresentable?

adlrocha replied:
Definitely. The reason why I added the `UndefState` is because in Golang we used it to signal an illegal state, but in the Rust case we can remove the `UndefState` and start the enum with `Initialized = 1`. Will push the change in the next commit.

aakoshh commented on actors/hierarchical_sca/src/exec.rs:
```diff
@@ -0,0 +1,197 @@
+/// Data persisted in the SCA for the orchestration of atomic executions.
+#[derive(Clone, Debug, PartialEq, Eq, Serialize_tuple, Deserialize_tuple)]
+pub struct AtomicExec {
+    pub params: AtomicExecParams,
+    pub submitted: HashMap<String, Cid>,
+    pub status: ExecStatus,
```
Maybe these fields could be private, with methods to access them (like `pub fn status(&self) -> ExecStatus`), and mutation only allowed by methods that take `&mut self`?

adlrocha replied:
Fixed!

aakoshh commented on actors/hierarchical_sca/src/exec.rs:
```diff
@@ -0,0 +1,197 @@
+    pub fn new(params: AtomicExecParams) -> Self {
+        AtomicExec {
+            params,
+            submitted: HashMap::<String, Cid>::new(),
```
Can we use a newtype or at least a type alias for what this `String` stands for?

aakoshh commented on actors/hierarchical_sca/src/exec.rs:
```diff
@@ -0,0 +1,197 @@
+/// Parameters used to submit the result of an atomic execution.
+#[derive(Clone, PartialEq, Eq, Serialize_tuple, Deserialize_tuple)]
+pub struct SubmitExecParams {
+    pub cid: Cid,
```
It would be helpful to have docs to say what it's a CID of.

aakoshh commented on actors/hierarchical_sca/src/exec.rs:
```diff
@@ -0,0 +1,197 @@
+#[derive(Clone, PartialEq, Eq, Serialize_tuple, Deserialize_tuple)]
+pub struct SubmitExecParams {
+    pub cid: Cid,
+    pub abort: bool,
+    pub output: atomic::SerializedState, // TODO: LockedState
```
Are these mutually exclusive? It would be great to use an enum that is either abort, or an output. Is that problematic with CBOR? 

Should it say what output it expects if it's an abort? Does it have to be empty vector, or can it be absolutely anything?

adlrocha replied:
> Are these mutually exclusive? It would be great to use an enum that is either abort, or an output. Is that problematic with CBOR?

I thought about that, but while I don't think there will be a problem in the CBOR serialization, it may be hard to implement the interoperability with Rust, that is why I decided to leave it as-is for now (it looked simpler).

> Should it say what output it expects if it's an abort? Does it have to be empty vector, or can it be absolutely anything?

We completely disregard the output if it's an abort. The default is an empty vector, but you can set whatever you want there (although I wouldn't expect this to happen, as the bigger the message, the more gas it consumes)

aakoshh replied:
Okay so it is a feature to allow the aborter to send some kind of an output to the others, and we hope they will handle it with enough care not to be an attack vector.

aakoshh commented on actors/hierarchical_sca/src/exec.rs:
```diff
@@ -0,0 +1,197 @@
+#[derive(Clone, Debug, PartialEq, Eq, Serialize_tuple, Deserialize_tuple)]
+pub struct AtomicExecParams {
+    pub msgs: Vec<StorableMsg>,
+    pub inputs: HashMap<String, LockedStateInfo>,
```
What does the `String` key stand for?

aakoshh replied:
I see it's some kind of an address, a hierarchical one by the looks of it. Can we create a type for that? 
Does it have to be related to the `actor` inside `LockedStateInfo`?

adlrocha replied:
It's the address of the user involved in the execution. I added some clarifications and aliased the type for clarity.

aakoshh commented on actors/hierarchical_sca/src/exec.rs:
```diff
@@ -0,0 +1,197 @@
+impl Cbor for LockedStateInfo {}
+
+#[derive(PartialEq, Eq, Clone, Serialize_tuple, Deserialize_tuple)]
+pub struct MetaExec {
```
Would `AtomicExecParamsMeta` be a more suggestive name, or do the two have nothing to do with each other?

adlrocha replied:
Yes, good point. Fixed.

aakoshh commented on actors/hierarchical_sca/src/exec.rs:
```diff
@@ -0,0 +1,197 @@
+#[derive(PartialEq, Eq, Clone, Serialize_tuple, Deserialize_tuple)]
+pub struct MetaExec {
+    pub msgs_cid: TCid<TAmt<StorableMsg>>,
+    pub input_cid: TCid<THamt<Address, LockedStateInfo>>,
```
```suggestion
    pub inputs_cid: TCid<THamt<Address, LockedStateInfo>>,
```
Just because in `AtomicExecParams` it's `msgs` and `inputs`.

aakoshh commented on actors/hierarchical_sca/src/exec.rs:
```diff
@@ -0,0 +1,197 @@
+            };
+            // Update with id_addr and subnet
+            let sn_addr = Address::new_hierarchical(&sn, &id_addr)?;
+            out.insert(sn_addr.to_string(), (*val).clone());
```
I think this cloning could be avoided if you used something like `let inputs = mem::take(&mut self.inputs);`. Then `self.inputs` becomes empty and you can iterate through the result as you own it. 

Of course it also depends on what you want to happen in case there is an error, whether it would be okay to leave `self.inputs` empty if there was corrupt data in it to start with.

adlrocha replied:
I tried to do that, but I am afraid you need to clone it either way. As we consume the iterator the value is behind a shared reference (I may be doing it wrong, though).

aakoshh replied:
I tried, it works like this:
```rust
        let inputs = std::mem::take(&mut self.inputs);
        let mut out = HashMap::default();
        for (key, val) in inputs {
            let addr = Address::from_str(&key)?;
            let sn = addr.subnet()?;
            let addr = addr.raw_addr()?;
            let id_addr = match rt.resolve_address(&addr) {
                Some(id) => id,
                None => return Err(anyhow!("couldn't resolve id address in exec input")),
            };
            // Update with id_addr and subnet
            let sn_addr = Address::new_hierarchical(&sn, &id_addr)?;
            out.insert(sn_addr.to_string(), val);
        }
        self.inputs = out;
        Ok(())
```
Again if anything goes wrong it leaves `self.inputs` empty, but maybe that's acceptable in this scenario.

I was curious if it would work with a `map`, so just for reference here's another version:

```rust
       self.inputs = std::mem::take(&mut self.inputs)
            .into_iter()
            .map(|(key, val)| {
                let addr = Address::from_str(&key)?;
                let sn = addr.subnet()?;
                let addr = addr.raw_addr()?;
                let id_addr = match rt.resolve_address(&addr) {
                    Some(id) => id,
                    None => return Err(anyhow!("couldn't resolve id address in exec input")),
                };
                // Update with id_addr and subnet
                let sn_addr = Address::new_hierarchical(&sn, &id_addr)?;
                Ok((sn_addr.to_string(), val))
            })
            .collect::<Result<HashMap<_, _>, _>>()?;
        Ok(())
```

aakoshh replied:
Another option to consider would be to consume `self` rather than take `&mut self` and return it in the result on success. That way it's obvious that we don't think there's a meaningful way to recover from having wrong addresses in the input, and we can do partial moves. 

But the ultimate solution would be to have a different type altogether for the resolved addresses with not `String` but something that shows that this step has not been forgotten, as I have mentioned elsewhere.

aakoshh replied:
This could be achieved by making `AtomicExecParams` generic in the key type, so in the actor method parameters it could be `params: AtomicExecParams<String>`, but in the state of `AtomicExec` it could be `pub params: AtomicExecParams<HierarchicalActorAddress>` or whatever name is best.

And then `impl AtomicExecParams<String>` could have the `input_into_ids` method that returns `Result<AtomicExecParams<HierarchicalActorAddress>>` instead of `Self`.

aakoshh replied:
Here's a PR with that change: https://github.com/adlrocha/builtin-actors/pull/15

I apologize for intruding on your PR, just wanted to see how it would work and didn't want to you to suffer with something that might be rubbish. Feel free to ignore it if you're happy with using `Address` and `String`.

adlrocha replied:
I really need to learn how to use idiomatic Rust, I love your approach. And by all means, feel free to jump in my PRs and even directly commit to the current branch.

adlrocha replied:
Merged the PR, feel free to make any additional changes directly over the PR.

adlrocha replied:
OK, I am afraid this PR broke the tests somehwere, did they worked for you? I'll try to fix them before reverting the commit.

adlrocha replied:
I think we are facing the reason why we were using `String` as key in the first place. Addresses seem to not be supported as keys in a map (at least this was the case in Go, and thus the use of strings). I may revert the PR for now so I can keep addressing your comments.

aakoshh commented on actors/hierarchical_sca/src/exec.rs:
```diff
@@ -0,0 +1,197 @@
+            msgs_array.batch_set(self.msgs.clone()).map_err(|e| e.into())
+        })?;
+
+        for (k, v) in self.inputs.iter() {
+            meta.input_cid.update(&store, |input_map| {
```
Could this two swap places, so not every intermediate map is stored individually?

adlrocha replied:
I don't think I understand what you mean here. Would you mind extending a bit? Thanks

aakoshh replied:
Of course, sorry for being cryptic. What I mean is that we take each `(k, v)` pair in `inputs`, deserialise the `meta.input_cid` from the `store` as `input_map`, insert `(k, v)`, serialise and persist the modified `input_map` and store the resulting CID in `input_cid`. So if we have `n items in `inputs` we will have `n` versions of the map in `store`, but we only need the CID of the last one. 

If, instead, we put the loop inside the `update`, we can get away with just retrieving/storing the map once:
```rust
        meta.input_cid.update(&store, |input_map| {
            for (k, v) in self.inputs.iter() {
                let addr = Address::from_str(k)?;
                input_map.set(addr.to_bytes().into(), v.clone()).map_err(|e| {
                    e.downcast_wrap(format!("failed to set input map to compute exec cid"))
                })?;
            }
            Ok(())
        })?;
```

adlrocha replied:
Oh man, you are right! I don't know why I did it that way. Fixed.

aakoshh commented on actors/hierarchical_sca/src/exec.rs:
```diff
@@ -0,0 +1,197 @@
+    inputs: &HashMap<String, LockedStateInfo>,
+) -> anyhow::Result<bool> {
+    let ks: Vec<String> = inputs.clone().into_keys().collect();
+    let addr = Address::from_str(ks[0].as_str())?;
```
Are you absolutely certain that `inputs` will never be empty, and `[0]` cause a panic?

aakoshh replied:
I see that you can because you call this after checking. It would be nice if the code didn't assume this, though.

adlrocha replied:
Fixed.

aakoshh commented on actors/hierarchical_sca/src/exec.rs:
```diff
@@ -0,0 +1,197 @@
+    curr: &SubnetID,
+    inputs: &HashMap<String, LockedStateInfo>,
+) -> anyhow::Result<bool> {
+    let ks: Vec<String> = inputs.clone().into_keys().collect();
```
Do we really have to clone the whole input map just to get the keys?

aakoshh commented on actors/hierarchical_sca/src/exec.rs:
```diff
@@ -0,0 +1,197 @@
+    caller: &Address,
+    inputs: &HashMap<String, LockedStateInfo>,
+) -> anyhow::Result<bool> {
+    let ks: Vec<String> = inputs.clone().into_keys().collect();
```
Can you use [.keys()](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.keys) to avoid cloning the whole map? Perhaps `inputs.keys().map(to_addr).find(|a| a == caller).is_some()` for the whole thing?

adlrocha replied:
Yes, the clone was not needed (I don't know why I did it that way). Fixed.

aakoshh commented on actors/hierarchical_sca/src/lib.rs:
```diff
@@ -629,6 +635,245 @@ impl Actor {
+                ));
+            }
+            None => {
+                // check if exec has correct number of inputs and messages.
+                if params.msgs.len() == 0 || params.inputs.len() < 2 {
+                    return Err(actor_error!(
+                        illegal_argument,
+                        "wrong number of messages or inputs provided for execution"
+                    ));
+                }
```
```suggestion
                // check if exec has correct number of inputs and messages.
                if params.inputs.len() < 2 {
                    return Err(actor_error!(
                        illegal_argument,
                        "wrong number of messages or inputs provided for execution"
                    ));
                }
```
It looks redundant. Maybe the comment or the error message should say what it expected. 
As I understand "input" needs to correspond to the number of participants, which should be at least 2.

aakoshh replied:
Oh it's `msgs` not `input`, sorry :see_no_evil:

aakoshh commented on actors/hierarchical_sca/src/lib.rs:
```diff
@@ -629,6 +635,245 @@ impl Actor {
+                    ));
+                }
+
+                // TODO: check if the atomic execution is initiated in the same address for different
+                // subnets? (that would be kind of stupid -.-)
```
What kind of addresses can we deal with here? As I understand there's no guarantee that `t01234` in one subnet is the entity in a different subnet. 

If it was a public key, could atomic swap be used as some kind of a synchronisation primitive in a sharded application that exists across networks?

adlrocha replied:
These are always key addresses, no ID addresses supported. We check that the caller is signable, but we should probably also check that the inputs provided are of the right type.

> If it was a public key, could atomic swap be used as some kind of a synchronisation primitive in a sharded application that exists across networks?

That's a great point, then it may not be as stupid to set two different inputs in different subnets with the same case (we may need some slight changes to support this).

aakoshh commented on actors/hierarchical_sca/src/lib.rs:
```diff
@@ -629,6 +635,245 @@ impl Actor {
+
+                // sanity-check: verify that all messages have same method and are directed to the same actor
+                // NOTE: This can probably be relaxed in the future
+                let method = params.msgs[0].method;
```
I don't see any checks before this to prevent a panic if `msgs` is empty.

adlrocha replied:
Yes, it is checked at the beginning: https://github.com/adlrocha/builtin-actors/blob/39e2f60c8a18ce481a91cdc8738d6a2b9a7c21d6/actors/hierarchical_sca/src/lib.rs#L681 but worth checking :)

aakoshh commented on actors/hierarchical_sca/src/lib.rs:
```diff
@@ -629,6 +635,245 @@ impl Actor {
+                }
+                Some(mut exec) => {
+                    // check if the output is aborted or already succeeded
+                    if exec.status != ExecStatus::Initialized {
+                        return Err(actor_error!(
+                            illegal_state,
+                            format!("execution with cid {} doesn't exist", &cid)
+                        ));
+                    }
+
+                    // check if the user is involved in the execution
+                    if !is_addr_in_exec(&caller, &exec.params.inputs).map_err(|e| {
+                        e.downcast_default(
+                            ExitCode::USR_ILLEGAL_ARGUMENT,
+                            "error checking if address is involved in the execution",
+                        )
+                    })? {
+                        return Err(actor_error!(
+                            illegal_argument,
+                            format!("caller not part of the execution for cid {}", &cid)
+                        ));
+                    }
```
These checks could be moved into the `exec` itself, mapping to custom error types that could then have methods to be turned into `ActorError` (or just return `ActorError` if easier). Maybe easier to test than in the big handler that needs a full runtime.

adlrocha replied:
Definitely, I will a TODO to defer it to the next PR.

adlrocha replied:
Actually, I added an issue to track this: https://github.com/adlrocha/builtin-actors/issues/14

aakoshh commented on actors/hierarchical_sca/src/lib.rs:
```diff
@@ -629,6 +635,245 @@ impl Actor {
+                        st.propagate_exec_result(
+                            rt.store(),
+                            &exec,
+                            params.output,
```
What output can the aborter send to all subnets? How can anyone trust whatever the content is?

adlrocha replied:
The output will only be accepted if all the participants have sent the same output (i.e. the same cid). Now that you mention this, I think there may be a potential attack here where a set of users can collude to push an invalid state to a subnet through a successful atomic execution, but I don't think this is possible if at least one of the participants is honest.

I'll add an issue describing the potential attack to keep it in mind for the future.

adlrocha replied:
Tracking issue here: https://github.com/filecoin-project/eudico/issues/228

aakoshh commented on actors/hierarchical_sca/src/lib.rs:
```diff
@@ -629,6 +635,245 @@ impl Actor {
+                    if exec.status != ExecStatus::Initialized {
+                        return Err(actor_error!(
+                            illegal_state,
+                            format!("execution with cid {} doesn't exist", &cid)
```
```suggestion
                            format!("execution with cid {} no longer accepts submissions", &cid)
```

aakoshh commented on actors/hierarchical_sca/src/lib.rs:
```diff
@@ -629,6 +635,245 @@ impl Actor {
+                        )
+                    })?;
+
+                    // TODO: Clean the execution once is done? Then we need
+                    // to traverse several epochs if we want to have a commands
+                    // that prompts the user with the state of all the execution
```
Sorry, I don't understand this sentence :thinking:

aakoshh commented on actors/hierarchical_sca/src/lib.rs:
```diff
@@ -629,6 +635,245 @@ impl Actor {
+                    // that prompts the user with the state of all the execution
+                    // (after succeeding or aborting). If we don't do this we need
+                    // a way to de-duplicate executions, because a user may abort an
+                    // execution and look to make it again (use the epoch?).
```
Wouldn't they have to send new `msgs`? Does a `msg` include a nonce, signature?

aakoshh commented on actors/hierarchical_sca/src/exec.rs:
```diff
@@ -0,0 +1,197 @@
+use anyhow::anyhow;
```
Should `exec.rs` be in the `atomic` directory?

adlrocha replied:
I was planning for the `atomic` crate to only include the atomic primitives used by actor to support atomic executions, and for `exec.rs` to include atomic-specific functions required by SCA. If you are not against this, let's decide in the next PR, once we finish implementing all the primitives and we have a sample actor using them.

aakoshh replied:
I'm not against anything, just think that anything that isn't strictly usable by the SCA should not be in the `hierarchical_sca` crate. The data types in the `atomic` package are targeted for implementors of those interfaces, but as far as I can tell, the SCA only cares about it as much as it uses the method IDs; everything else is opaque data. To put iit differently, why should an implementor of an actor that supports atomic execution reference the whole `hierarchical_sca` just to get these types?

In any case, everything in `exec` is still really `atomic_exec`, so IMO belongs in an atomic namespace. 

Of course it can wait, it's just my opinion.

adlrocha replied:
> To put iit differently, why should an implementor of an actor that supports atomic execution reference the whole hierarchical_sca just to get these types?

Yes, sure. The atomic primitives need to go in its own namespace (and potentially each own repo, or the SDK repo). And now I see what you mean regarding `exec`. I'll fix it.

aakoshh commented on actors/hierarchical_sca/src/lib.rs:
```diff
@@ -629,6 +635,245 @@ impl Actor {
+                        )
+                    })?;
+
+                    // TODO: Clean the execution once is done? Then we need
```
I think clearing out this state is a good idea. 

In the beginning, the participants already know from off-chain communication what each others' locked CIDs are, and they agree on the messages they will send. All of them have a copy, so anyone (even non-participants) can initialise the execution. Then the participants send their outputs, which is really just a signature proving that they all agree to carry out the execution. The state is resolved off-chain, on-chain only a CID appears, and the messages. Once everyone agrees, the parent subnet sends down a message to the subnets to commit to the results that they individually calculated. By the virtue of every child subnet following the parent, they will all eventually deliver this message and commit.

Keeping the tombstone around in the parent doesn't seem to serve a purpose any more. In theory anyone could look at past executions and replicate the results, if they can resolve all input state and have access to the off-line actor, but they can find out the same thing by replaying messages from the chain, and this doesn't seem to be part of the regular atomic execution flow, so why should the parent be responsible for carrying this state around?

adlrocha replied:
I am convinced and you are totally right. I will add this, so when an execution succeed or is aborted it is cleaned from the registry.

aakoshh commented on actors/hierarchical_sca/src/state.rs:
```diff
@@ -478,6 +479,116 @@ impl State {
+        &mut self,
+        store: &BS,
+        exec: &AtomicExec,
+        output: atomic::SerializedState, /* FIXME: LockedState to propagate. The same as in SubmitAtomicExecParams*/
```
Yeah I don't think you can have a generic parameter here, because you are executing in an arbitrary common ancestor where you have no idea what kind of state was locked in some descendant subnet's actor that wants to do an atomic execution.

On a different note, `LockedState` has that `lock: bool` flag in it. Is that supposed to be `true` for the _output_?

adlrocha replied:
> Yeah I don't think you can have a generic parameter here, because you are executing in an arbitrary common ancestor where you have no idea what kind of state was locked in some descendant subnet's actor that wants to do an atomic execution.

Agree.

> On a different note, LockedState has that lock: bool flag in it. Is that supposed to be true for the output?

Correct, and is only unlocked once it has been merged into the state of the subnet through the top-down message.

aakoshh commented on actors/hierarchical_sca/src/state.rs:
```diff
@@ -478,6 +479,116 @@ impl State {
+        let ks: Vec<String> = exec.params.inputs.clone().into_keys().collect();
+        let mut visited = HashMap::<SubnetID, bool>::new();
+        for k in ks.iter() {
+            let addr = Address::from_str(k.as_str())?;
+            let sn = addr.subnet()?;
```
Since we do this every time, maybe a custom type that has the subnet and the raw address already resolved could be used as a key, constructed during `init_atomic_exec`? There's already `input_into_ids` that mutates the keys, so it seems like we're constantly second guessing our own computation and entertaining the possibility that the input is bad even when the computation is concluded.

aakoshh replied:
For things to fail at this stage would be a disaster, leaving state locked in the subnets, so ideally there should be as little legal opportunities for it as possible.

adlrocha replied:
This issue is addressed with #15, right?

aakoshh commented on actors/hierarchical_sca/src/state.rs:
```diff
@@ -478,6 +479,116 @@ impl State {
+                    let mut msg = self.exec_result_msg(
+                        &sn,
+                        &input.actor,
+                        &exec.params.msgs[0],
```
Why send `msgs[0]` in particular, what's special about that one?

adlrocha replied:
All the messages have to be directed to the same method and the same address in the current implementation (meaning we can only send messages to the same actor for the atomic execution --actually we explicitly check this--). As the destination and the method are the same for all the messages, it's OK for us to just pick the first one to create the top-down message that will be propagated down. We should probably extend the protocol to support an arbitrary number of methods within the same actor, but we wanted to test if the protocol made sense before complicating it.

aakoshh replied:
What I don't understand is that while it's true that `msgs` are checked for `to` and and `method` are the same, the way `exec_result_msg` uses it is this:
```
let lock_params = atomic::LockParams::new(msg.method, msg.clone().params);
```
It uses `params` which is _not_ checked to be the same across all `msgs`. What guarantees that this will be a valid `lock_params` instance from every participant's point of view?

adlrocha replied:
In `LockedStateInfo` we include information about the cid of the locked state and the actor where it is locked. When we call `propagate_exec_resul` we set the destination for each subnet to the right actor specified in the `LockedStateInfo`. This of course assumes that only one actor can be involved in the atomic execution. See:  https://github.com/adlrocha/builtin-actors/blob/f64a12606a303ad373ac13bcf549ce4cd6fbd05e/actors/hierarchical_sca/src/state.rs#L541

aakoshh replied:
> This of course assumes that only one actor can be involved in the atomic execution

Exactly, that's why I suggested that `visited` should include `(subnet, actor)` not just `subnet`, so that if they are different, all of them get notified. The other option is to verify that assumption in the init and reject if it doesn't hold.

I still haven't heard why `params` would be the same across all `msgs`.

aakoshh commented on actors/hierarchical_sca/src/state.rs:
```diff
@@ -478,6 +479,116 @@ impl State {
+            let addr = Address::from_str(k.as_str())?;
+            let sn = addr.subnet()?;
+            match visited.get(&sn) {
+                Some(_) => {
+                    continue;
+                }
```
Why skip? I understand some actors could live in the same subnet, but the parameters to `exec_result_msg` vary by `k` (ie. `input.actor` is different), so it seems arbitrary to use one and not the other.

adlrocha replied:
We skip the subnets that we have already visited. We propagate a single cross-net message with all the atomic execution information required to merge the output and unlock the state (so if we have several addresses in the same subnet we just need a single top-down message).

aakoshh replied:
Okay, but these messages would not necessarily be identical, because they are constructed like so:
```
 let mut msg =
                        self.exec_result_msg(&sn, &input.actor, &p.msgs[0], output.clone(), abort)?;
                    self.send_cross(store, &mut msg, curr_epoch)?;
```
The `input.actor` is just the value from the first input from that subnet. How do we know it's not two different actors that want to participate in an atomic swap, and both should get their own messages? 

If `actor` has to be the same across all inputs from the same subnet, then it should be checked during initialisation. If not, then maybe `visited` should contain not just `sn` but the `actor` as well, to make sure we're not missing anyone.

adlrocha replied:
I think this is the related with the other comment, as part of the locking stage we need to specify the source ID of the actor involved in the execution in `LockedStateInfo`. This is later used to propagate the commit message and propagate the unlock message to the right place.

aakoshh commented on actors/hierarchical_sca/src/state.rs:
```diff
@@ -478,6 +479,116 @@ impl State {
+                }
+                None => {
+                    // send cross-message
+                    let input = match exec.params.inputs.get(k) {
+                        Some(i) => i,
+                        None => {
+                            return Err(anyhow!(
+                                "input for subnet not found. unable to propagate the output message"
+                            ))
+                        }
+                    };
```
We are iterating over the keys of `exec.params.input` so it would be reason to `panic` if the value was not in the map. But to resist all temptation for handling this error, why not iterate through keys and values at the same time?

adlrocha replied:
You're totally right. I was translating from Go and I ended up with this horrible thing. Fixed!

aakoshh commented on actors/hierarchical_sca/src/state.rs:
```diff
@@ -478,6 +479,116 @@ impl State {
+                    };
+                    let mut msg = self.exec_result_msg(
+                        &sn,
+                        &input.actor,
```
What if the `actor` value is wrong? I haven't seen it checked during initialisation. It would mean that some participants never get notifications about results, right? 

So what about someone trying to attack like this: Alice, Bob, Charlie and Eve want to participate in a 4 way atomic transaction. Everyone locks their state and give the input CID to Eve. Eve, initialises the atomic swap at the closest common ancestor, but corrupts everyone's `actor` field except herself. Then she also initialises another swap correctly. 

Before the others could submit their outputs, Eve _aborts_ the first atomic swap - since hers is the only correct `actor` in it, only her subnet gets properly notified; the others go into the dead letters. This notification unlocks her state, but the others don't know this, and they go ahead with the 2nd atomic swap. Eve gets a second notification about the successful swap, but she's already in a state where her state isn't locked. 

It seems useless to do this, but the point is that the parent system and the child system are both individually correct, Eve is managing to put them in unexpected states.

Maybe there should be some check that all locked states are only participating in a single atomic swap at any time, in the parent?

adlrocha replied:
Ok! I think you are right and this could happen. Even more, it may be also worth to add a check to see if all the actors involved have the same code CID (as if this is not the case they may lead to the wrong output and inconsistent states). I'll include it in the attack-vector issue mentioned above.

aakoshh commented on actors/hierarchical_sca/src/state.rs:
```diff
@@ -478,6 +479,116 @@ impl State {
+                        &sn,
+                        &input.actor,
+                        &exec.params.msgs[0],
+                        output.clone(),
```
How does the receiving system know which input, which atomic execution has been completed here?

Sorry, I should probably know by now but `output` could be anything, since `LockedState` is generic, isn't it?

adlrocha replied:
Yes, that is why we use `LockParams` and `UnlockParams` to signal the type of state locked and the output type. I hope this makes more sense once I am done with the sample actor supporting atomic execution.

aakoshh commented on actors/hierarchical_sca/src/state.rs:
```diff
@@ -478,6 +479,116 @@ impl State {
+        // to signal that is a system message we use system_actor_addr as source.
+        let from = Address::new_hierarchical(&self.network_name, &SYSTEM_ACTOR_ADDR)?;
+        let to = Address::new_hierarchical(subnet, actor)?;
+        let lock_params = atomic::LockParams::new(msg.method, msg.clone().params);
```
Is this supposed to identify the thing to unlock? But we pass `msgs[0]` which could be originally from a different actor. I at least thought that `params` could vary depending on who sends them, and there can be unlimited amount of messages as well. Is everyone supposed to treat the first one special? What if the first message is the same across multiple atomic executions?

aakoshh commented on actors/hierarchical_sca/src/lib.rs:
```diff
@@ -629,6 +635,245 @@ impl Actor {
+    {
+        rt.validate_immediate_caller_type(CALLER_TYPES_SIGNABLE.iter())?;
+        let caller = rt.message().caller();
+        let mut out_status = ExecStatus::UndefState;
```
If `UndefState` exists only to allow you to declare this variable, then I think Rust has your back here, you can just say `let out_status: ExecStatus;` and the compiler will make sure that you initialise it on all conditional branches, like [here](https://github.com/ChainSafe/forest/blob/a5b5cf616a66106b2e1caa4adfda183ead0b21a7/forest/src/daemon.rs#L213).

aakoshh commented on actors/hierarchical_sca/src/exec.rs:
```diff
@@ -0,0 +1,232 @@
+            let addr = Address::from_str(&key)?;
+            let sn = addr.subnet()?;
+            let addr = addr.raw_addr()?;
+            let id_addr = match rt.resolve_address(&addr) {
```
This means that atomic execution doesn't work if any of the participants don't have an account in the common parent subnet, ie. if their address is not recognised and cannot be turned into an ID. I didn't realise this before, it looks like a bit of a snag. 

Where is the ID format used? I tried looking for all places where `inputs` are iterated over but all I could see was accessing the `subnet()`.

aakoshh replied:
The only place where I can see something is might have an effect is in `is_addr_in_exec` where `caller` could be of a different type. But maybe `caller` could be resolved to a key instead?

It would be awesome if there was at least some phantom type to say what kind of address we can be dealing with. Or maybe a _newtype_ wrapper that rejects stuff during deserialisation if it's not the expected format. It could enhance our confidence in doing comparisons by a lot if we didn't have to wonder whether `caller` is a BLS key or an ID, and whether the atomic execution parameters have or haven't been turned into IDs. Internally it could still be just `Address` (not an enum), but we could do away with a lot of `?` if we knew when we can or cannot call `.subnet()` for example.

adlrocha replied:
> This means that atomic execution doesn't work if any of the participants don't have an account in the common parent subnet, ie. if their address is not recognised and cannot be turned into an ID. I didn't realise this before, it looks like a bit of a snag.

We'll need to test this once we perform the integration of the actor in the node. The initial assumption is that if the use has some state in the subnet it potentially injected some funds there (which means that it already has an address in the parent). However, if this is not the case, the node can initialize a new ID for every address involved in the atomic execution if it doesn't exist. This will be handled by the peer, so we need to defer that to the end-to-end integration.

aakoshh replied:
Okay, so the host system is expected to do some preparations. I'll try to remember that, but back to the second question, why do we have to resolve keys to IDs at all?

adlrocha replied:
Actors only understand on-chain ID addresses. When a message is sent to an actor, `rt.caller()` returns the ID of the address that signed the message. To avoid having to resolve every time one of the parties interact with the actor, we store internally the map with their on-chain IDs.

aakoshh replied:
Right, its' the caller then. Cool, so I should be able to use `TAddress<ID>` for that one.

aakoshh commented on actors/hierarchical_sca/src/state.rs:
```diff
@@ -478,6 +479,118 @@ impl State {
+        curr_epoch: ChainEpoch,
+        abort: bool,
+    ) -> anyhow::Result<()> {
+        let mut visited = HashMap::<SubnetID, bool>::new();
```
Sorry just noticing: you can use [HashSet](https://doc.rust-lang.org/std/collections/struct.HashSet.html).

aakoshh commented:
Thanks for following through with all my comments. 

I hope that the next PR with the lockable actor will make things clearer, but I was wondering: is the call to `init_atomic_exec` of the SCA coming from a user directly? Or is it coming as a bottom-up message via checkpoints? 

I'm curious how the system can ensure that the same state is not locked in multiple atomic executions that have different levels as their common ancestor. Say we have subnets like this:

```
     S0
    /  \ 
   S1  S2
  /  \
S3    S4
```
And there's an actor in S3 that wants to do an atomic swap with one in S4 and another in S2. It locks some state. Then what prevents the user from sending the same locked CID to both S1 and S0? Or sending something that isn't locked at all?

If the "I locked some state for atomic execution X" actually comes from the actor in S3 then it's fine because it can only lock once. But that's not the impression I get from the spec.

Ultimately I want guarantee that if something is locked it's locked, if somebody is committed to the output they are committed, and I can only see this guaranteed if the messages are unforgeable and come from the actors involved, not crafted by users. 

You might say that the users can go to each others' subnets and check with the respective actors whether the state is locked. But looking at the model, all I can see is in the `LockedMapCid` is that the state they claimed was locked is indeed locked _now_, but there's no guarantee it's locked because of the atomic execution I am interested in. The next block might unlock them if they also locked it for a different atomic execution that I am unaware of and it's already finishing, and then I'll be committed to something they no longer are, and when the output from our atomic execution reaches them they'll just drop it.

aakoshh commented:
Just theorizing, but maybe the atomic execution protocol could go like this:
1. Participants call `lock` on some state of theirs to get the corresponding input CID.
2. They all agree on the `AtomicExecutionParams` from the `msgs` and `inputs` they now know. 
3. One of them sends the `AtomicExecutionParams` to the nearest common ancestor, they all learn the resulting CID of the record - although it's a good question how they can trust anyone telling them what it is because of changing the address strings to IDs. Maybe the CID should be that of the original, raw params?
4. _New step:_ Everyone calls `pre_commit(lock_cid, atomic_exec_params_cid, output)` _in their own subnet_. The lockable actor now associates the locked data with this execution, and sends the commitment to `output_cid` up the chain as a bottom-up message. 
5. The SCA in the atomic execution subnet gathers to pre-commits, and issues the `commit(atomic_exec_params_cid)` once everyone's vote has arrived. 
6. The actors in the descendant subnets receive the commit message and do the unlock.

So, the change are that the `output` isn't coming from a top-down message, it's the participants entering it into their respective subnets; doing so triggers a pre-commit message going up, and the locked state being associated with the atomic execution, so it can't be associated with anything else, until `abort` or `commit` arrives.

aakoshh commented:
@adlrocha I see that this force-push undid the changes in my PR: https://github.com/adlrocha/builtin-actors/compare/12d67b98bc7c144b8956a0db119a9792cfa06651..a24c90e249da694a1ef35723b94d26e2479a0155

Not sure if it was intentional or you just forgot to pull?

adlrocha commented:
> @adlrocha I see that this force-push undid the changes in my PR: https://github.com/adlrocha/builtin-actors/compare/12d67b98bc7c144b8956a0db119a9792cfa06651..a24c90e249da694a1ef35723b94d26e2479a0155
> Not sure if it was intentional or you just forgot to pull?

It was intentional, sorry. Using addresses broke something in the tests and I had to revert the commit so I could keep addressing your comments (see https://github.com/adlrocha/builtin-actors/pull/13#discussion_r930110428). Did tests passed for you? I actually just tested your branch again and it fails, and I am afraid that this is because `Address` is not supported as a key for maps for the CBOR serialization (this is why I used `Strings` in the first place :( ).

adlrocha commented:
> I hope that the next PR with the lockable actor will make things clearer, but I was wondering: is the call to init_atomic_exec of the SCA coming from a user directly? Or is it coming as a bottom-up message via checkpoints?

In the current implementation it is the user the one that directly sends a message in the common parent to initialize or submit the execution, now cross-net messages involved.

> I'm curious how the system can ensure that the same state is not locked in multiple atomic executions that have different levels as their common ancestor.

You are totally right. We wouldn't be able right now to identify that a user is using the same locked state for two executions. Actually, in the current implementation we wouldn't even be able to identify that the same locked state is used in two computations in the same common parent (we are not doing that check, and we should). 

I opened [this discussion](https://github.com/protocol/ConsensusLab/discussions/154) to continue there (I don't want comments to get lost in this PR).

aakoshh commented:
> and I am afraid that this is because Address is not supported as a key for maps for the CBOR serialization (this is why I used Strings in the first place :(

How odd. Then maybe we can wrap it with yet another type, an `AddressKey` type, that overrides the `Cbor` serialisation to use the `String` format of the wrapped address.

Sorry, I didn't run the tests, they don't even compile for me because some cases in a match aren't handled. Let me give it another shot then.

adlrocha commented:
> Sorry, I didn't run the tests, they don't even compile for me because some cases in a match aren't handled. Let me give it another shot then.

I've moved a lot of things around, so you may need to rebase if we want to be able to merge the change directly, I am really sorry :( I am starting to work on the sample actor supporting atomic execution so I won't touch anything from this PR so you can do your magic. Thanks :pray:

aakoshh commented:
Okay, so I'm still not 100% convinced about these:
* https://github.com/adlrocha/builtin-actors/pull/13#discussion_r930284319
* https://github.com/adlrocha/builtin-actors/pull/13#discussion_r930287000

But if you think these are okay, then merge ahead :)

adlrocha commented:
> Okay, so I'm still not 100% convinced about these:
> 
>     * [Atomic execution protocol and primitives #13 (comment)](https://github.com/adlrocha/builtin-actors/pull/13#discussion_r930284319)
> 
>     * [Atomic execution protocol and primitives #13 (comment)](https://github.com/adlrocha/builtin-actors/pull/13#discussion_r930287000)
> 
> 
> But if you think these are okay, then merge ahead :)

Answered, let me know if it makes sense. In any case, as we are proposing some changes to the overall protocol, maybe it's a good idea to have this merged so we have the core to start including the new changes (once we have the whole design).

adlrocha commented:
Quick update on this: I think that with the recent discussions we're having, the basic atomic primitives will change a lot, and we will be able to lock several parts of the state using different messages over the same actor. The new architecture should remove many of the simplifications used in the MVP. Parts of the code from this PR will change a bit with this new approach. Let's note down our concerns, merge this PR and review them in the next one. WDYT?

aakoshh commented:
I agree, let's merge it.

aakoshh reviewed (approved):
I agree, let's merge this PR, it was a great platform to discuss, but we already know some things will change so there's no reason to keep arguing about the minute details here.
